scalableminds/webknossos
# Dependency-Injecting DAOs and dependents

This PR is a big step towards upgrading Play. It
 - significantly reduces the (deprecated) usage of the global state of the framework `Play.current`
 - reduces the usage of `GlobalSettings`
 - …by creating most Services and DAOs as dependency-injected classes
 - it removes circular dependencies amongst these

 -  note that not all of the newly injected classes are initialized `asEagerSingleton` – we might add that as a further optimization, but I don’t want to make this PR any bigger right now.
 - Nothing can be injected into case classes, so they can’t provide much functionality anymore, stuff like `annotation.task` has been moved to `annotationService.taskFor(annotation)`. Same with `publicWrites` methods
 - since silhouette is now injected into the controllers, the implicit conversions from SecuredRequest to DBAccessContext can’t be in a trait anymore, which leads to a bit of code duplication, but I think we can live with that for now
 - the resolution of cyclic dependencies brought some refactoring, for example we no longer need the explicit conversion `securedRequestToUserAwareRequest` (we pass on the Option[User] rather than the request)
 - there are still eight occurances of Play.current, which block advancing to 2.6 together with the streaming stuff (and a possibly required silhouette upgrade). We will adress those in future PRs

### TODO
 - [x] remove currently commented-out `ProvidesUnauthorizedSessionData` (I believe it is unused or fully replaced by silhouette, since the tests run through and manual testing didn’t show changed behavior, I don’t think we need it anymore)
 - [x] ensureTeamAdministration should be replaced by `assertTrue(userService.isTeamManagerOrAdminOf)` – the `assertTrue` is missing from some places
 - [x] `annotationService.zippedFourLists` should be renamed

### Steps to test:
 - stuff the CI doesn’t cover:
 - reload after scala code changes should close the db handle + open a new one (check console output)
 - Mailer should still work
 - cleanup services should still work (try reducing the interval for testing)

### Issues:
- fixes #3059 
- contributes to #3102 
(which are both part of #1622 )

------
~~- [ ] Needs datastore update after deployment~~
